File: python/src/azoth/eos/reference/vs_flash.py
"""``eos.vs_flash`` - the pressure and temperature a mixture reaches at a given volume and
entropy.

Spec: ``specs/models/eos/vu_flash.toml``

A closed vessel at fixed volume and entropy: both state variables are solved for,
by the decoupled 2x2 Newton in :mod:`azoth.eos.reference._flash_property`.
"""

from __future__ import annotations

from azoth import _models_gen
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import VsFlashResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.eos.mixture import Mixture
from azoth.eos.reference._flash_property import distinct_warnings, solve_pressure_temperature
from azoth.eos.reference.molar_enthalpy_entropy import IdealGasModel
from azoth.eos.reference.pr_molar_volume import MOLAR_GAS_CONSTANT

MODEL_ID = "eos.vs_flash"


def vs_flash(
    mixture: Mixture,
    ideal_gas: IdealGasModel,
    V: Q,
    S: Q,
    z: list[float],
) -> VsFlashResult:
    """The pressure and temperature at which a mixture has a given volume and entropy.

    Raises:
        OutOfRangeError: if ``V`` is not positive.
        SolverNotConvergedError: if the iteration reaches its cap.
    """
    spec = _models_gen.model(MODEL_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    v_si = input_to_si(spec, "V", V)
    s_si = input_to_si(spec, "S", S)
    apply_checks(checks.on_input, {"V": v_si, "S": s_si}.get, warnings)

    start_t = float(spec["algorithm"].get("initial_temperature", 300.0))
    start_p = MOLAR_GAS_CONSTANT * start_t / v_si
    solved = solve_pressure_temperature(
        mixture,
        ideal_gas,
        v_si,
        ("s", s_si),
        list(z),
        spec["algorithm"],
        start_p,
        start_t,
    )
    warnings.extend(solved["warnings"])
    flash = solved["state"]["flash"]

    return VsFlashResult(
        P=from_si(solved["P"], "Pa"),
        T=from_si(solved["T"], "K"),
        beta=flash.beta,
        x=tuple(flash.x),
        y=tuple(flash.y),
        k=tuple(flash.k),
        phase=flash.phase,
        z_liquid=flash.z_liquid,
        z_vapour=flash.z_vapour,
        iterations=solved["iterations"],
        residual=solved["residual"],
        warnings=distinct_warnings(warnings),
    )

File: crates/azoth-eos/src/vs_flash.rs
//! `eos.vs_flash` - the pressure and temperature a mixture reaches at a given volume and
//! entropy.
//!
//! Spec: `specs/models/eos/vs_flash.toml`
//!
//! A vessel at fixed volume and entropy: both state variables are solved for, by the same
//! decoupled 2x2 Newton [`crate::vu_flash`] uses, with an entropy asked for instead of an
//! energy - NeqSim's `VSflash`, which is its `OptimizedVUflash` loop with a Q function
//! in the entropy rather than in the enthalpy.

use azoth_core::units::{MolarHeatCapacity, MolarVolume, kelvins, pascals};
use azoth_core::{Result, apply_checks};

use crate::algorithm_of;
use crate::flash_property::{EnergyTarget, solve_pressure_temperature};
use crate::mixture::Mixture;
use crate::model_gen;
use crate::molar_enthalpy_entropy::IdealGasModel;
use crate::pr_molar_volume::MOLAR_GAS_CONSTANT;
use crate::results::VsFlashResult;

/// The pressure and temperature at which a mixture has a given molar volume and entropy.
///
/// # Errors
/// * [`azoth_core::AzothError::OutOfRange`] if `v` is not positive.
/// * [`azoth_core::AzothError::SolverNotConverged`] if the iteration reaches its cap.
pub fn vs_flash(
    mixture: &Mixture,
    ideal_gas: &IdealGasModel,
    v: MolarVolume,
    s: MolarHeatCapacity,
    z: &[f64],
) -> Result<VsFlashResult> {
    let spec = &model_gen::VS_FLASH_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "V" => Some(v.value),
            "S" => Some(s.value),
            _ => None,
        },
        &mut warnings,
    )?;

    let algorithm = algorithm_of(spec)?;
    let start_t = algorithm.initial_temperature.unwrap_or(300.0);
    let start_p = MOLAR_GAS_CONSTANT * start_t / v.value;
    let solved = solve_pressure_temperature(
        mixture,
        ideal_gas,
        v.value,
        EnergyTarget::Entropy(s.value),
        z,
        algorithm,
        start_p,
        start_t,
    )?;

    Ok(VsFlashResult {
        pressure: pascals(solved.pressure),
        temperature: kelvins(solved.temperature),
        beta: solved.flash.beta,
        x: solved.flash.x,
        y: solved.flash.y,
        k: solved.flash.k,
        phase: solved.flash.phase,
        z_liquid: solved.flash.z_liquid,
        z_vapour: solved.flash.z_vapour,
        iterations: solved.iterations,
        residual: solved.residual,
        warnings: solved.warnings,
    })
}
